README.md

```diff
@@ -21,3 +21,65 @@ While testing against a real kafka has many benefits for integration testing, it
+
+In the case of the repartition messages, we want them to cause a repartition in our system mock of the partitions, so we catch them and feed them back into the system using the `CompletingInternalsDriver`, which allows them to be separated onto different partitions by the `PartitioningDriver`.
+
+In the case of the source messages, we want them to get the opportunity to be shuffled and read in a different order, so we catch them and feed them back into the system using the `RecursionDriver`, which allows them to be reordered by the `ShuffleDriver` 
+
+The `TopologyDriver` then returns a map of topic name (for non-repartition topics) or topic properties map (for repartition topics) to a list of messages published to that topic.
+
+### The PartitioningDriver
+
+This driver (in the `partitioned-driver` namespace) maintains a collection of `TopologyDriver`s and a fn (defined through the options argument but set by default to be the message key round-tripped through the serialiser) which generates a mock partition key. Every time the `PartitioningDriver` receives a new message, it forwards it to the `TopologyDriver` corresponding to the message's key, creating a new one if it has never seen that key before.
+
+The idea is that this simulates the partitions being on separate nodes, since each message key corresponds to a different driver, so will be processed entirely independently.
+
+### The CompletingInternalsDriver
+
+This driver wraps a `PartitioningDriver`. When it receives a new message, it passes it along to the `PartitioningDriver`. It then feeds any repartitioning messages in the output back into the `PartitioningDriver`, repeating the process until there is no more output. Along the way it collects any non-repartitioning output, and returns it at the end. This allows the `PartitioningDriver` to allocate work to a new "partition" whenever the system switches to a new partition with a repartitioning message.
+
+### The CombiningDriver
+
+This driver can be found in the `combined-driver` namespace. It receives a collection of name-value pairs of topologies, and maintains a `CompletingInternalsDriver` for each one. Every time it's asked to publish a message, it publishes it to each topology's `CompletingInternalsDriver`, and every time it returns messages, it amalgamates the results from each topology's `CompletingInternalsDriver`. This allows us to simulate a system with multiple topologies rather than just one.
+
+### The ApplyingSerdeDriver
+
+This driver (in the `serde-driver` namespace) allows clients to deal with standard clojure maps. It delegates to a `CombiningDriver`, serialising all messages on the way in, and deserialising all messages on the way out.
```

Comment:
I have a hunch that this driver isn't doing much anymore since the upgrade....

Reply:
Ah yeah! I wrote the docs against an earlier version and must have missed that when I checked the commits since. Have removed this from the docs.